src/vs/workbench/contrib/chat/browser/chatContentParts/toolInvocationParts/chatTerminalToolProgressPart.ts

```diff
@@ -67,6 +70,200 @@ const sanitizerConfig = Object.freeze<DomSanitizerConfig>({
  */
 const expandedStateByInvocation = new WeakMap<IChatToolInvocation | IChatToolInvocationSerialized, boolean>();
 
+interface ITerminalCommandDecorationOptions {
+	readonly terminalData: IChatTerminalToolInvocationData;
+	getTitleElement(): HTMLElement | undefined;
+	getHost(): HTMLElement | undefined;
+	getResolvedCommand(): ITerminalCommand | undefined;
+}
+
+class TerminalCommandDecoration extends Disposable {
+	private _element: HTMLElement | undefined;
+	private _interval: number | undefined;
+
+	constructor(private readonly _options: ITerminalCommandDecorationOptions) {
+		super();
+		this._register(toDisposable(() => this._stopTimer()));
+	}
+
+	public ensureElement(context?: HTMLElement): HTMLElement | undefined {
+		const title = context ?? this._options.getTitleElement();
+		if (!title) {
+			return undefined;
+		}
+		// eslint-disable-next-line no-restricted-syntax
+		const markdown = title.querySelector<HTMLElement>('.rendered-markdown');
+		const parent = markdown?.parentElement;
+		if (!markdown || !parent) {
+			return undefined;
+		}
+
+		let container = parent;
+		if (!container.classList.contains('chat-terminal-command-block')) {
+			const wrapper = document.createElement('div');
+			wrapper.classList.add('chat-terminal-command-block');
+			container.insertBefore(wrapper, markdown);
+			wrapper.appendChild(markdown);
+			container = wrapper;
+		}
+		// eslint-disable-next-line no-restricted-syntax
+		let decoration = container.querySelector<HTMLElement>(':scope > .chat-terminal-command-decoration');
+		if (!decoration) {
+			decoration = document.createElement('span');
+			decoration.classList.add('chat-terminal-command-decoration');
+			decoration.setAttribute('role', 'img');
+			container.insertBefore(decoration, markdown);
+		}
+
+		this._element = decoration;
+		return decoration;
+	}
+
+	public update(command?: ITerminalCommand): void {
+		const decoration = this.ensureElement();
+		if (!decoration) {
+			this._stopTimer();
+			return;
+		}
+		const resolvedCommand = command ?? this._options.getResolvedCommand();
+		this._apply(decoration, resolvedCommand);
+		const storedState = this._options.terminalData.terminalCommandState;
+		if (resolvedCommand?.exitCode === undefined || (!resolvedCommand && storedState && storedState.exitCode === undefined)) {
+			this._startTimer();
```

Comment:
Considered keeping track of duration on the `ITerminalCommand` instead. LMK if that's preferable.